config/config-mp/src/main/java/io/helidon/config/mp/MpConfigImpl.java

```diff
@@ -365,32 +386,70 @@ private String resolveReferences(String key, String value) {
         }
         if (!UNRESOLVED_KEYS.get().add(key)) {
             UNRESOLVED_KEYS.get().clear();
-            throw new IllegalStateException("Recursive resolving of references for key " + key + ", value: " + value);
+            throw new IllegalArgumentException("Recursive resolving of references for key " + key + ", value: " + value);
         }
         try {
-            return format(value);
-        } catch (NoSuchElementException e) {
-            LOGGER.log(Level.FINER, e, () -> String.format("Reference for key %s not found. Value: %s", key, value));
-            return value;
+            if (value.contains("${")) {
+                return processExpressions(value);
+            } else {
+                return value;
+            }
```

Comment:
You don't seem to like conditional expressions :)

Reply:
Yes. I always found them trickier to read.

Reply:
Nevertheless fixed, as this one is really simple